community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -0,0 +1,481 @@
+        KEY key = layout.newKey();
+        VALUE value = layout.newValue();
+        cursor.next();
+
+        boolean isInternal;
+        int keyCount = 0; // initialized to satisfy compiler
+        long childId = 0; // initialized to satisfy compiler
+        int pos;
+        do
+        {
+            do
+            {
+                isInternal = bTreeNode.isInternal( cursor );
+                // Find the left-most key within from-range
+                keyCount = bTreeNode.keyCount( cursor );
+                pos = 0;
+                int search = KeySearch.search( cursor, bTreeNode, fromInclusive, key, keyCount );
+                pos = KeySearch.positionOf( search );
+
+                // TODO: if we don't have unique keys we cannot make this assumption,
+                // but instead we'd need to go to the previous child
+                if ( isInternal && KeySearch.isHit( search ) )
+                {
+                    pos++;
+                }
+
+                if ( isInternal )
+                {
+                    childId = bTreeNode.childAt( cursor, pos, stableGeneration, unstableGeneration );
+                }
+            }
+            while ( cursor.shouldRetry() );
+            cursor.checkAndClearCursorException();
+            if ( cursor.checkAndClearBoundsFlag() )
+            {
+                // Something's wrong, get a new fresh rootId and go from the top again
+                cursor.next( localRootId = rootId );
+                continue;
+            }
+            if ( isInternal )
+            {
+                // TODO perhaps detect being too far to the left and so follow right siblings,
+                // to reduce unnecessary seek in leafs (concurrent splits would be the cause)
+
+                PointerChecking.checkChildPointer( childId );
+
+                if ( !cursor.next( childId ) )
+                {
+                    throw new IllegalStateException( "Couldn't go to child " + childId );
+                }
+            }
+        }
+        while ( isInternal && keyCount > 0 );
+
+        // Returns cursor which is now initiated with left-most leaf node for the specified range
+        return new SeekCursor<>( cursor, key, value, bTreeNode, fromInclusive, toExclusive, layout,
+                stableGeneration, unstableGeneration, pos, keyCount );
+    }
+
+    @Override
+    public long acquireNewId()
+    {
+        lastId++;
+        return lastId;
+    }
+
+    // Utility method
+    public void printTree() throws IOException
+    {
+        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
+        {
+            cursor.next();
+            TreePrinter.printTree( cursor, bTreeNode, layout, stableGeneration, unstableGeneration, System.out );
+        }
+    }
+
+    // Utility method
+    public boolean consistencyCheck() throws IOException
+    {
+        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
```

Comment:
Yes, definitely. Suggest postponing this work for a different PR, as we need to refactor entire ConsistencyChecker to not be recursive. Card created.